File: day_3/part_1.cpp
#include <iostream>
#include <fstream>
#include <string> 

using namespace std;

vector<vector<int>> create_rows(){
    ifstream inputFile("input.txt");
    vector<vector<int>> rows;
    string line;
    while( !inputFile.eof() ) {
        getline( inputFile, line );
        vector<int> nums;
        for (char n: line) {
            nums.push_back(n - '0');
        }
        rows.push_back(nums);
    }
    
    return rows;
}
int main() {
    vector<vector<int>> rows = create_rows();
    int res = 0;
    for(vector<int> row : rows) {
        int l1[2] = {0, -1};
        int l2[2] = {0, -1};
        for (int i = 0; i < row.size() - 1; i++) {
            if (l1[0] < row[i]) {
                l1[0] = row[i];
                l1[1] = i;
            }
        }
        for (int i = row.size() - 1; i > l1[1]; i--) {
            if (l2[0] < row[i]) {
                l2[0] = row[i];
            }
        }
        int num = (l1[0] * 10) + l2[0];
        res += num;
    }
    cout << res;
    return res;
};

File: day_3/part_1.py
def solve(inp):
    rows = []
    with open(inp, "r") as file:
        lines = file.readlines()
        for l in lines:
            rows.append(list(map(lambda x: int(x), l.replace("\n", ""))))

    res = 0
    for row in rows:
        l1, l2 = (0, -1), (0, -1)
        for i in range(len(row) - 1):
            l1 = max(l1, (row[i], i), key=lambda x: (x[0], -x[1]))
        for i in range(len(row) - 1, l1[1], -1):
            l2 = max(l2, (row[i], i))
        res += int(f"{l1[0]}{l2[0]}")
    return res


# print(solve("example.txt"))
print(solve("input.txt"))
